Dart/src/com/jetbrains/lang/dart/analyzer/DartAnalysisServerColorAnnotator.java

```diff
@@ -0,0 +1,147 @@
+    }
+    if (type.equals(HighlightRegionType.GETTER_DECLARATION)) {
+      return DartSyntaxHighlighterColors.DART_GETTER_DECLARATION;
+    }
+    if (type.equals(HighlightRegionType.LOCAL_VARIABLE)) {
+      return DartSyntaxHighlighterColors.DART_LOCAL_VARIABLE_ACCESS;
+    }
+    if (type.equals(HighlightRegionType.IMPORT_PREFIX)) {
+      return DartSyntaxHighlighterColors.DART_IMPORT_PREFIX;
+    }
+    if (type.equals(HighlightRegionType.KEYWORD)) {
+      return DartSyntaxHighlighterColors.DART_KEYWORD;
+    }
+    if (type.equals(HighlightRegionType.LABEL)) {
+      return DartSyntaxHighlighterColors.DART_LABEL;
+    }
+    if (type.equals(HighlightRegionType.LOCAL_VARIABLE_DECLARATION)) {
+      return DartSyntaxHighlighterColors.DART_LOCAL_VARIABLE;
+    }
+    if (type.equals(HighlightRegionType.METHOD)) {
+      return DartSyntaxHighlighterColors.DART_INSTANCE_MEMBER_FUNCTION_CALL;
+    }
+    if (type.equals(HighlightRegionType.METHOD_STATIC)) {
+      return DartSyntaxHighlighterColors.DART_STATIC_MEMBER_FUNCTION_CALL;
+    }
+    if (type.equals(HighlightRegionType.METHOD_DECLARATION)) {
+      return DartSyntaxHighlighterColors.DART_INSTANCE_MEMBER_FUNCTION;
+    }
+    if (type.equals(HighlightRegionType.METHOD_DECLARATION_STATIC)) {
+      return DartSyntaxHighlighterColors.DART_STATIC_MEMBER_FUNCTION;
+    }
+    if (type.equals(HighlightRegionType.PARAMETER)) {
+      return DartSyntaxHighlighterColors.DART_PARAMETER;
+    }
+    if (type.equals(HighlightRegionType.SETTER_DECLARATION)) {
+      return DartSyntaxHighlighterColors.DART_SETTER_DECLARATION;
+    }
+    if (type.equals(HighlightRegionType.TOP_LEVEL_VARIABLE)) {
+      return DartSyntaxHighlighterColors.DART_TOP_LEVEL_VARIABLE_DECLARATION;
+    }
+    if (type.equals(HighlightRegionType.TYPE_NAME_DYNAMIC)) {
+      return DartSyntaxHighlighterColors.DART_KEYWORD;
+    }
+    if (type.equals(HighlightRegionType.TYPE_PARAMETER)) {
+      return DartSyntaxHighlighterColors.DART_TYPE_PARAMETER;
+    }
+    return null;
+  }
+
+  @Override
+  public void apply(@NotNull PsiFile file, Object annotationResult, @NotNull AnnotationHolder holder) {
+    final VirtualFile virtualFile = file.getVirtualFile();
+    final List<PluginHighlightRegion> regions = DartAnalysisServerService.getInstance().getHighlight(virtualFile);
+    for (PluginHighlightRegion region : regions) {
+      final String type = region.getType();
+      final String key = getHighlightType(type);
+      if (key != null) {
+        final TextAttributesKey textAttributes = TextAttributesKey.createTextAttributesKey(key);
+        final TextRange textRange = new TextRange(region.getOffset(), region.getOffset() + region.getLength());
+        holder.createInfoAnnotation(textRange, null).setTextAttributes(textAttributes);
+      }
+    }
+  }
+
+  @Nullable
+  @Override
+  public Object collectInformation(@NotNull PsiFile file, @NotNull Editor editor, boolean hasErrors) {
```

Comment:
I'm thinking about combining DartAnalysisServerColorAnnotator with DartAnalysisServerAnnotator.
collectInformation() will be as in DartAnalysisServerAnnotator: it performs checks and sends files updated content.
doAnnotate() will wait both for errors and for updated highlighting data for the current file.
In this case we'll get correct file highlighting in one cycle of highlighting daemon.